diff --git a/src/internal/client/dom/blocks/boundary.js b/src/internal/client/dom/blocks/boundary.js
--- a/src/internal/client/dom/blocks/boundary.js
+++ b/src/internal/client/dom/blocks/boundary.js
@@ -1,5 +1,5 @@
 /** @import { Effect, Source, TemplateNode, } from '#client' */
-import { BOUNDARY_EFFECT, EFFECT_PRESERVED, EFFECT_TRANSPARENT } from '#client/constants';
+import { BOUNDARY_EFFECT, DESTROYED, EFFECT_PRESERVED, EFFECT_TRANSPARENT } from '#client/constants';
 import { HYDRATION_START_ELSE, HYDRATION_START_FAILED } from '../../../../constants.js';
 import { component_context, set_component_context } from '../../context.js';
 import { invoke_error_boundary } from '../../error-handling.js';
@@ -427,12 +427,12 @@
 				});
 			}
 
-			if (this.#offscreen_fragment) {
+			if (this.#offscreen_fragment && (this.#effect.f & DESTROYED) === 0) {
 				var pop_renderer = push_renderer(this.#effect.r);
 				insert_before(this.#anchor, this.#offscreen_fragment);
-				this.#offscreen_fragment = null;
 				pop_renderer?.();
 			}
+			this.#offscreen_fragment = null;
 		}
 	}
 
